File: display.py
import os
import subprocess as sp
import time

"""
--------------------------------------------------
Display - V.1.0 - Written by Joshua Kitchen 5/21/2020
--------------------------------------------------
"""


class Display:
    def __init__(self, rows, columns, background=" "):
        self.columns = columns
        self.row_count = rows
        self.rows = [[] for _ in range(self.row_count)]
        self.x_movement = "right"
        self.y_movement = "up"
        self.offset_counter = 0
        self.background = background

        for i in range(self.columns):
            for row in self.rows:
                row.append(self.background)

    @staticmethod
    def clear_terminal():
        if os.name == 'nt':
            sp.call("cls", shell=True)
        else:
            sp.call("clear", shell=True)

    def draw(self, posy, posx, char):
        try:
            self.rows[posy][posx] = char
        except IndexError:
            print(f'Coordinates Y: {posy} X: {posx} are out of bounds.\n'
                  f'Display size is {len(self.rows)} X {self.columns}')
            return

    def clear_display(self):
        self.clear_terminal()
        self.rows = []
        self.rows = [[] for _ in range(self.row_count)]
        for i in range(self.columns):
            for row in self.rows:
                row.append(self.background)

    def show(self):
        for row in self.rows:
            print(*row)
            
    # Examples: animating text

    def up_and_down(self, text, speed, frames, justify='center'):
        if len(text) > self.columns:
            print(f"Selected String cannot fit in the current display size.\n"
                  f"Size of string: {len(text)}\n"
                  f"Size of display: 5 rows X {self.columns} columns")
            return
        
        self.clear_display()
        self.y_movement = "up"
        coords = []
        y = len(self.rows) / 2
        y = int(round(y, 1))
        if justify == 'left':
            x = 0
        elif justify == 'right':
            x = self.columns - len(text)
        elif justify == 'center':
            x = (self.columns - len(text)) / 2
            x = int(round(x, 1))
        else:
            print(f"Invalid argument: {justify}. Must be \'center\', \'left\' or \'right\'")
            return
        for char in text:
            self.draw(y, x, char)
            coords.append([y, x, char])
            x += 1

        self.show()

        count = 0
        while count < frames:
            time.sleep(speed)
            self.clear_display()
            if coords[0][0] == 0:
                self.y_movement = "down"
            elif coords[0][0] == self.row_count - 1:
                self.y_movement = "up"

            for i in range(len(coords)):
                if i > self.offset_counter:
                    pass
                else:
                    if self.y_movement == "up":
                        coords[i][0] = coords[i][0] - 1
                    elif self.y_movement == "down":
                        coords[i][0] = coords[i][0] + 1
                    self.draw(coords[i][0], coords[i][1], coords[i][2])
                self.offset_counter += 1
            self.show()
            count += 1

    def side_to_side(self, text, speed, frames, start_pos='center'):
        if len(text) > self.columns:
            print(f"Selected String cannot fit in the current display size.\n"
                  f"Size of string: {len(text)}\n"
                  f"Size of display: 5 rows X {self.columns} columns")
            return
        
        self.clear_display()
        self.x_movement = "left"
        coords = []
        if start_pos == 'top':
            y = 0
        elif start_pos == 'bottom':
            y = self.row_count - 1
        elif start_pos == 'center':
            y = len(self.rows) / 2
            y = int(round(y, 1))
        else:
            print(f"Invalid argument: {start_pos}. Must be \'center\', \'top\' or \'bottom\'")
            return
        x = (self.columns - len(text)) / 2
        x = int(round(x, 1))
        for char in text:
            self.draw(y, x, char)
            coords.append([y, x, char])
            x += 1
        self.show()

        count = 0
        while count < frames:
            time.sleep(speed)
            self.clear_display()
            if coords[0][1] == 0:
                self.x_movement = "right"
            elif coords[-1][1] == self.columns - 1:
                self.x_movement = "left"
            if self.x_movement == "right":
                for i in range(len(coords)):
                    coords[i][1] = coords[i][1] + 1
                    self.draw(coords[i][0], coords[i][1], coords[i][2])
            elif self.x_movement == "left":
                for i in range(len(coords)):
                    coords[i][1] = coords[i][1] - 1
                    self.draw(coords[i][0], coords[i][1], coords[i][2])
            self.show()
            count += 1

    def wavy_text(self, text, speed, frames):
        if len(text) > self.columns:
            print(f"Selected String cannot fit in the current display size.\n"
                  f"Size of string: {len(text)}\n"
                  f"Size of display: 5 rows X {self.columns} columns")
            return
        
        self.clear_display()
        self.offset_counter = 1
        coords = []
        y = len(self.rows) / 2
        y = int(round(y, 1))
        x = (self.columns - len(text)) / 2
        x = int(round(x, 1))
        for char in text:
            self.draw(y, x, char)
            coords.append([y, x, char, "up"])
            """                         ^   
                        direction of character (up or down"""
            x += 1
        self.show()

        count = 0
        while frames > count:
            time.sleep(speed)
            self.clear_display()
            for i in range(len(coords)):
                if i < self.offset_counter:
                    if coords[i][0] == 0:
                        coords[i][3] = "down"
                    elif coords[i][0] == self.row_count - 1:
                        coords[i][3] = "up"
                    if coords[i][3] == "up":
                        coords[i][0] = coords[i][0] - 1
                    elif coords[i][3] == "down":
                        coords[i][0] = coords[i][0] + 1
                else:
                    pass

                self.draw(coords[i][0], coords[i][1], coords[i][2])
            self.offset_counter += 1
            self.show()
            count += 1

    def scroll_in_horiz(self, text, speed, stop=None):
        if len(text) > self.columns:
            print(f"Selected String cannot fit in the current display size.\n"
                  f"Size of string: {len(text)}\n"
                  f"Size of display: 5 rows X {self.columns} columns")
            return
        
        self.clear_display()
        self.x_movement = "right"
        if stop is None:
            stop = self.columns - 1
        else:
            if stop > self.columns - 1:
                print(f'Stopping point: {stop} is greater than column count: {self.columns - 1}')
        coords = []
        y = len(self.rows) / 2
        y = int(round(y, 1))
        for char in text:
            coords.append([None, char])
            """           ^   
                        x_pos """
        self.offset_counter = -2
        self.show()
        while True:
            try:
                if coords[0][0] > stop:
                    break
            except TypeError:
                pass
            time.sleep(speed)
            self.clear_display()
            i = -1
            for _ in range(len(coords)):
                if i > self.offset_counter:
                    if coords[i][0] is None:
                        coords[i][0] = 0
                    else:
                        coords[i][0] += 1
                    if coords[i][0] >= self.columns:
                        pass
                    else:
                        self.draw(y, coords[i][0], coords[i][1])
                else:
                    pass
                i -= 1
            self.offset_counter -= 1
            self.show()

    def scroll_in_vert(self, text, speed, stop=None, justify='center'):
        if len(text) > self.columns:
            print(f"Selected String cannot fit in the current display size.\n"
                  f"Size of string: {len(text)}\n"
                  f"Size of display: 5 rows X {self.columns} columns")
            return
        
        self.clear_display()
        self.x_movement = "right"
        self.offset_counter = 0
        if stop is None:
            stop = self.row_count - 1
        coords = []
        if justify == 'left':
            x = 0
        elif justify == 'right':
            x = self.columns - len(text)
        elif justify == 'center':
            x = (self.columns - len(text)) / 2
            x = int(round(x, 1))
        else:
            print(f"Invalid argument: {justify}. Must be \'center\', \'left\' or \'right\'")
            return
        for char in text:
            coords.append([0, x, char])
            x += 1
        self.show()
        while True:
            if coords[0][0] == stop:
                break
            time.sleep(speed)
            
            self.clear_display()
            for i in range(len(coords)):
                coords[i][0] = coords[i][0] + 1
                self.draw(coords[i][0], coords[i][1], coords[i][2])
            self.show()


if __name__ == '__main__':
    screen = Display(7, 20)
    screen.scroll_in_horiz('Hello World', 0.1)
    screen.scroll_in_vert('Hello World', 0.1, 3)
    time.sleep(0.2)
    screen.side_to_side('Hello World', 0.1, 18)
    screen.up_and_down('Hello World', 0.1, 12)
    screen.wavy_text('Hello World', 0.1, 35)

File: display.cpp
/* 
* Display (C++ version)
* Written by Joshua Kitchen 3/27/2021
*
* Error codes:
* 1 = out of bounds
* 2 = bad argument
*/
#include <iostream>
#include <cstdlib>
#include <vector>
#include <chrono>
#include <thread>

class Pixel{
    public:
        int x;
        int y;
        char value;

        Pixel(int pixel_x, int pixel_y, char pixel_value){
            x = pixel_x;
            y = pixel_y;
            value = pixel_value;
        }
};


class Sprite{
    public:
    std::vector<Pixel> pixels;
    int length;
    int height;

        Sprite(std::vector<Pixel> sprite_pixels){
            pixels = sprite_pixels;
            int largest_x = pixels[0].x;
            int largest_y = pixels[0].y;

            int smallest_x = pixels[0].x;
            int smallest_y = pixels[0].y;

            for(Pixel p:pixels){
                if(p.x > largest_x){
                    largest_x = p.x;
                }
                if(p.x < smallest_x){
                    smallest_x = p.x;
                }

                if(p.y > largest_y){
                    largest_y = p.y;
                }
                if(p.y < smallest_y){
                    smallest_y = p.y;
                }
            }

            length = (largest_x - smallest_x) + 1;
            height = (largest_y - smallest_y) + 1;
        }

        void show_coords(){
            for(Pixel p:pixels){
                std::cout << p.value << ": (" << p.x << ", " << p.y << ")\n"; 
            }
        }

        // direction: left, right, up, down
        // length: how far to move the sprite
        int move(std::string direction, int distance){
            for(int i=0;i<=pixels.size() - 1;i++){
                if(direction ==  "left"){
                    pixels[i].x = pixels[i].x - distance;
                }
                else if(direction == "right"){
                    pixels[i].x = pixels[i].x + distance;
                }
                else if(direction == "up"){
                    pixels[i].y = pixels[i].y - distance;
                }
                else if(direction == "down"){
                    pixels[i].y = pixels[i].y + distance;
                }
                else{
                    return 2;
                }
            }
            return 0;
        }      
};


class Display{

    int rows;
    int columns;
    char bg;
    std::vector<std::vector<char>> screen;

    public:
        Display(int display_rows, int display_columns, char display_bg){
            rows = display_rows;
            columns = display_columns;
            bg = display_bg;
            std::vector<char> temp;
            for(int r=0; r<=rows-1; r++){
                temp.clear();
                for(int c=0;c<=columns-1;c++){
                    temp.push_back(bg);
                }
                screen.push_back(temp);
            }
        }

        void clear_terminal(){
            std::system("clear");
        }

        void clear_display(){
            for(int r=0; r<=rows-1; r++){
                for(int c=0; c<=columns-1;c++){
                    screen[r][c] = bg;
                }
            }
        }

        int length(){
            return columns;
        }

        int height(){
            return rows;
        }

        void change_bg(char new_bg){
            for(int r=0; r<=rows-1; r++){
                for(int c=0; c<=columns-1;c++){
                    char value = screen[r][c];
                    if(value == bg){
                        screen[r][c] = new_bg;
                    }
                }
            }
            bg = new_bg;
        }

        int draw_char(int x_pos, int y_pos, char character){
            if(x_pos > columns - 1 or x_pos < 0){
                return 1;
            }

            if(y_pos > rows - 1 or y_pos < 0){
                return 1;
            }

            screen[y_pos][x_pos] = character;
            return 0;
        }

        std::vector<int> draw_sprite(Sprite sprite){
            std::vector<int> errorCodes;
            for(Pixel p:sprite.pixels){
                int exit_code = draw_char(p.x, p.y, p.value);
                errorCodes.push_back(exit_code);
            }
            return errorCodes;
        }

        void show(bool show_grid=false){
            for(int r=rows-1; r>=0; r--){
                if(show_grid == true){
                    if(r < 10){
                        std::cout << r << "  ";
                    }
                    else{
                        std::cout << r << " ";
                    } 
                }
                for(int c=0; c<=columns - 1;c++){
                    if(show_grid == true){
                        std::cout << screen[r][c] << " ";
                    }
                    else{
                        std::cout << screen[r][c];
                    }
                }
                std::cout << "\n";
            }

            if(show_grid == true){
                int column_counter = 0;
                for(int i=0; i<=columns + 2; i++){
                    if(i <= 2){
                        std::cout << " ";
                    }
                    else{
                        if(column_counter < 10){
                            std::cout << column_counter << " ";
                        }
                        else{
                            std::cout << column_counter;
                        }
                        column_counter ++;
                    }
                }
                std::cout << "\n";
            }
        }
        
};


void scrolling_text(Display display_obj, std::string text, int speed=100, int stop=-1, std::string direction="rightToLeft"){

    int y = display_obj.height() / 2;
    std::vector<int> char_coords;
    std::vector<char> char_vector;
    for(char s:text){
        if(direction == "leftToRight"){
            char_coords.push_back(-1);
        }
        else if(direction == "rightToLeft"){
            char_coords.push_back(display_obj.length());
        }
        else{
            std::cout << "Bad argument " << "\"" << direction << "\"" << " for direction" << "\n";
            return;
        }
        
        char_vector.push_back(s);
    }

    if(direction == "leftToRight"){
        if(stop < 0){
        stop = display_obj.length() - 1;
        }

        int offset = char_vector.size() + 1;
        while(char_coords[0] <= stop){
            std::this_thread::sleep_for(std::chrono::milliseconds(speed));
            display_obj.clear_display();
            display_obj.clear_terminal();

            for(int i=char_vector.size() - 1;i>=0;i--){
                if(i >= offset){
                    char_coords[i] ++;
                    if(char_coords[i] > display_obj.length() - 1){} // ignore if char is out of view
                    else{
                        display_obj.draw_char(char_coords[i], y, char_vector[i]);
                    }
                }
            }
            offset --;
            display_obj.show();
        }
    }
    else if (direction == "rightToLeft"){
        int offset = -1;
        while(char_coords.back() >= stop){
            std::this_thread::sleep_for(std::chrono::milliseconds(speed));
            display_obj.clear_display();
            display_obj.clear_terminal();

            for(int i=0;i<=char_vector.size() - 1;i++){
                if(i <= offset){
                    char_coords[i] --;
                    if(char_coords[i] < 0){} // ignore if char is out of view
                    else{
                        display_obj.draw_char(char_coords[i], y, char_vector[i]);
                    }
                }
            }
            offset ++;
            display_obj.show();
        }
    }
}


int main(int argc, char *argv[]){
    Display screen = Display(25, 51, '*');
    scrolling_text(screen, "Hello World");
}